### PR TITLE
[MXNET-729] Scala Examples memory leak fix

### PR DESCRIPTION
**Pull request by lanking520**
## Description ##
Currently the Scala integration test running are strong influenced by CUDA memory not enough. In order to address that issue, this PR gives a test run on the `NDArrayCollector` created by @yzhliu.
@andrewfayres @nswamy 

related CI failure:
http://jenkins.mxnet-ci.amazon-ml.com/blue/organizations/jenkins/incubator-mxnet/detail/PR-11753/9/pipeline

http://jenkins.mxnet-ci.amazon-ml.com/blue/organizations/jenkins/incubator-mxnet/detail/PR-11753/10/pipeline

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [x] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [x] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### CODE REVIEW

File: scala-package/examples/src/test/scala/org/apache/mxnetexamples/rnn/ExampleRNNSuite.scala [hunk 1]
```diff
@@ -50,26 +49,32 @@ class ExampleRNNSuite extends FunSuite with BeforeAndAfterAll {
       System.getenv("SCALA_TEST_ON_GPU").toInt == 1) {
       ctx = Context.gpu()
     }
-    LstmBucketing.runTraining(tempDirPath + "/RNN/sherlockholmes.train.txt",
+    NDArrayCollector.auto().withScope {
+      LstmBucketing.runTraining(tempDirPath + "/RNN/sherlockholmes.train.txt",
```
Comment (yzhliu) [inline scala-package/examples/src/test/scala/org/apache/mxnetexamples/rnn/ExampleRNNSuite.scala]: it is better to have the collector inside `runTraining`, like that you do in GanMnist.scala. Otherwise when the # of loop increases, too many ndarrays will be stored temporarily in the collector here.
But if you just mean to get rid of mem leak in CI, then this is fine.

Comment (lanking520) [reply scala-package/examples/src/test/scala/org/apache/mxnetexamples/rnn/ExampleRNNSuite.scala]: As you said, this piece of code is just used to improve the memory leak issues in the CI.

Comment (nswamy) [reply scala-package/examples/src/test/scala/org/apache/mxnetexamples/rnn/ExampleRNNSuite.scala]: @lanking520 Can you please make the change that Yizhi is asking? lets do it right because this will become a pattern in other parts of the code.

Comment (lanking520) [reply scala-package/examples/src/test/scala/org/apache/mxnetexamples/rnn/ExampleRNNSuite.scala]: @nswamy Depends on the purpose of the action:
1. Improving CI, this is the right action to wrap from outside
2. Improving model itself, then need to change from inside.

File: scala-package/examples/src/test/scala/org/apache/mxnetexamples/gan/GanExampleSuite.scala [hunk 2]
```diff
@@ -44,7 +44,9 @@ class GanExampleSuite extends FunSuite with BeforeAndAfterAll{
 
         val context = Context.gpu()
 
-        val output = GanMnist.runTraining(modelDirPath, context, modelDirPath, 5)
+        val output = NDArrayCollector.auto().withScope {
```
Comment (nswamy) [inline scala-package/examples/src/test/scala/org/apache/mxnetexamples/gan/GanExampleSuite.scala]: I don't think you need NDCollectors here. It should be sufficient to just have it inside the training loop(for each epoch)

**Comment by lanking520** [comment]
I only add a big giant {} in every example, not changing things inside.
